File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/Shortform/ShortformController.php
<?php

namespace App\Http\Controllers\Api\V1\Shortform;

use OpenApi\Attributes as OA;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\Shortform\Requests\StoreShortformCommentRequest;
use App\Modules\Shortform\Requests\StoreShortformRequest;
use App\Modules\Shortform\Services\ShortformService;
use Illuminate\Http\Request;

#[OA\Tag(name: 'Shortform')]
class ShortformController extends ApiController
{
    public function __construct(private readonly ShortformService $service) {}

    public function store(StoreShortformRequest $request)
    {
        return $this->success($this->service->create($request->user(), $request->validated()), [], 201);
    }

    public function index(Request $request)
    {
        return $this->success($this->service->list($request->query('cursor'), (int) $request->query('limit', 20)));
    }

    public function show(int $id)
    {
        return $this->success($this->service->show($id));
    }

    public function like(Request $request, int $id)
    {
        return $this->success($this->service->toggleLike($request->user(), $id));
    }

    public function storeComment(StoreShortformCommentRequest $request, int $id)
    {
        return $this->success($this->service->addComment($request->user(), $id, $request->validated()), [], 201);
    }

    public function comments(Request $request, int $id)
    {
        return $this->success($this->service->comments($id, (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function ranking(Request $request)
    {
        return $this->success($this->service->ranking((string) $request->query('period', 'day'), (int) $request->query('limit', 20)));
    }

    public function transcodeStatus(int $id)
    {
        return $this->success($this->service->transcodeStatus($id));
    }

    public function retry(Request $request, int $id)
    {
        return $this->success($this->service->retry($request->user(), $id));
    }

    public function destroy(Request $request, int $id)
    {
        return $this->success($this->service->delete($request->user(), $id));
    }

    public function userShortforms(Request $request, int $userId)
    {
        return $this->success($this->service->byUser($userId, (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/shortform/ShortformController.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.shortform

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.pbActor
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.respondStub
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.PbRole
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.requireAnyRole
import io.ktor.http.content.PartData
import io.ktor.http.content.forEachPart
import io.ktor.server.application.call
import io.ktor.server.request.receive
import io.ktor.server.request.receiveMultipart
import io.ktor.server.routing.Route
import io.ktor.server.routing.delete
import io.ktor.server.routing.get
import io.ktor.server.routing.post

class ShortformController(
    private val service: ShortformService,
) {
    fun Route.register() {
        post("/shortforms") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.upload(call.currentUserId(), call.readUploadInput()))
        }
        get("/shortforms/ranking/list") {
            call.respondStub(
                service.ranking(
                    period = call.request.queryParameters["period"],
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 10,
                ),
            )
        }
        get("/shortforms/user/{userId}") {
            call.respondStub(
                service.userShortforms(
                    userId = call.pathInt("userId"),
                    page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1,
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20,
                ),
            )
        }
        get("/shortforms/{id}/transcode-status") {
            call.respondStub(service.transcodeStatus(call.pathInt("id")))
        }
        post("/shortforms/{id}/transcode/retry") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.retry(call.currentUserId(), call.pbActor().role == PbRole.ADMIN, call.pathInt("id")))
        }
        post("/shortforms/{id}/like") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.toggleLike(call.currentUserId(), call.pathInt("id")))
        }
        post("/shortforms/{id}/comments") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.createComment(call.currentUserId(), call.pathInt("id"), call.receive()))
        }
        get("/shortforms/{id}/comments") {
            call.respondStub(
                service.comments(
                    shortformId = call.pathInt("id"),
                    page = call.request.queryParameters["page"]?.toIntOrNull() ?: 1,
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20,
                ),
            )
        }
        get("/shortforms/{id}") {
            call.respondStub(service.detail(call.pathInt("id")))
        }
        delete("/shortforms/{id}") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.delete(call.currentUserId(), call.pbActor().role == PbRole.ADMIN, call.pathInt("id")))
        }
        get("/shortforms") {
            call.respondStub(
                service.feed(
                    cursor = call.request.queryParameters["cursor"]?.toIntOrNull(),
                    limit = call.request.queryParameters["limit"]?.toIntOrNull() ?: 20,
                ),
            )
        }
    }

    private suspend fun io.ktor.server.application.ApplicationCall.readUploadInput(): ShortformUploadInput {
        val multipart = receiveMultipart()
        var title = ""
        var fileName: String? = null
        var mimeType = "video/mp4"
        val productIds = mutableListOf<Int>()
        multipart.forEachPart { part ->
            when (part) {
                is PartData.FileItem -> {
                    fileName = part.originalFileName ?: part.name ?: "shortform-video.mp4"
                    mimeType = part.contentType?.toString() ?: mimeType
                }
                is PartData.FormItem -> {
                    when (part.name) {
                        "title" -> title = part.value
                        "productIds", "productIds[]" -> part.value.toIntOrNull()?.let(productIds::add)
                    }
                }
                else -> Unit
            }
            part.dispose()
        }
        return ShortformUploadInput(
            title = title,
            originalFilename = fileName.orEmpty(),
            mimeType = mimeType,
            size = 1024,
            productIds = productIds,
        )
    }

    private fun io.ktor.server.application.ApplicationCall.currentUserId(): Int = pbActor().userId ?: 0

    private fun io.ktor.server.application.ApplicationCall.pathInt(name: String): Int = parameters[name]?.toIntOrNull() ?: 0
}
